sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/DelegateAnnotationVisitor.java

```diff
@@ -0,0 +1,251 @@
+import io.micronaut.sourcegen.model.VariableDef;
+import io.micronaut.sourcegen.model.VariableDef.This;
+
+import javax.lang.model.element.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * The visitor that is generation a delegate.
+ *
+ * @author Andriy Dmytruk
+ * @since 1.3
+ */
+@Internal
+public final class DelegateAnnotationVisitor implements TypeElementVisitor<Delegate, Object> {
+
+    public static final String DELEGATE_TYPE_MEMBER = "type";
+    public static final String DELEGATEE_MEMBER = "delegatee";
+    public static final String NAME_SUFFIX = "Delegate";
+
+    private final Set<String> processed = new HashSet<>();
+
+    @Override
+    public @NonNull VisitorKind getVisitorKind() {
+        return VisitorKind.ISOLATING;
+    }
+
+    @Override
+    public void start(VisitorContext visitorContext) {
+        processed.clear();
+    }
+
+    @Override
+    public Set<String> getSupportedAnnotationNames() {
+        return Set.of(Delegate.class.getName());
+    }
+
+    @Override
+    public void visitClass(ClassElement element, VisitorContext context) {
+        AnnotationValue<?> annotation = element.getAnnotation(Delegate.class);
+        Optional<Class<?>> type = annotation.classValue("type");
+
+        if (type.isPresent() && !type.get().equals(Void.class)) {
+            ClassElement typeElement = context.getClassElement(type.get()).orElseThrow(
+                () -> new ProcessingException(element, "Could not find required type " + type.get().getName()));
+            createDelegate(typeElement, context);
+        } else {
+            createDelegate(element, context);
+        }
+    }
+
+    private void createDelegate(ClassElement element, VisitorContext context) {
+        if (processed.contains(element.getName())) {
+            return;
+        }
+        if (!element.isInterface()) {
+            throw new ProcessingException(element, "Only interfaces are supported for delegate creation. But '"
+                + element.getName() + "' is annotated with @Delegate");
+        }
+
+        try {
+            String simpleName = element.getSimpleName() + NAME_SUFFIX;
+            String delegateClassName = element.getPackageName() + "." + simpleName;
+
+            ClassTypeDef typeDef = ClassTypeDef.of(element);
+            ClassTypeDef delegateType = ClassTypeDef.of(delegateClassName);
+            ClassDefBuilder delegate = ClassDef.builder(delegateClassName)
+                .addModifiers(Modifier.PUBLIC, Modifier.ABSTRACT);
+
+            if (!element.getTypeArguments().isEmpty()) {
+                for (Entry<String, ClassElement> type: element.getTypeArguments().entrySet()) {
+                    if (type.getValue() instanceof GenericPlaceholderElement placeholderElement) {
```

Comment:
Maybe create a new method `TypeVariable.of` to convert  `GenericPlaceholderElement` directly

Reply:
That's a good idea!